File: lightroom_tagger/core/database/stacks.py
"""Image stack membership and mutations."""

from __future__ import annotations

import sqlite3
from collections.abc import Sequence

from lightroom_tagger.core.exceptions import StackMutationError


def list_catalog_stack_member_keys(db: sqlite3.Connection, catalog_key: str) -> list[str]:
    """Return all ``image_key`` values in the stack containing *catalog_key*.

    If the key is not in ``image_stack_members``, returns ``[catalog_key]`` (solo image).
    """
    row = db.execute(
        "SELECT stack_id FROM image_stack_members WHERE image_key = ? LIMIT 1",
        (catalog_key,),
    ).fetchone()
    if not row:
        return [catalog_key]
    stack_id = int(row["stack_id"])
    rows = db.execute(
        "SELECT image_key FROM image_stack_members WHERE stack_id = ? ORDER BY image_key",
        (stack_id,),
    ).fetchall()
    return [str(r["image_key"]) for r in rows]


def select_stack_representative_key_for_keys(
    db: sqlite3.Connection, keys: Sequence[str]
) -> str | None:
    """Pick representative using the same ranking as ``batch_stack_detect`` (handlers)."""
    burst_keys = tuple(str(k) for k in keys if k)
    if not burst_keys:
        return None
    ph = ",".join("?" * len(burst_keys))
    sql = (
        "SELECT i.key AS k FROM images i "
        "LEFT JOIN ( "
        "  SELECT s.image_key AS image_key, AVG(s.score) AS ai_score "
        "  FROM image_scores s "
        "  INNER JOIN perspectives p ON p.slug = s.perspective_slug AND p.active = 1 "
        "  WHERE s.is_current = 1 AND s.image_type = 'catalog' "
        "  GROUP BY s.image_key "
        ") agg ON agg.image_key = i.key "
        f"WHERE i.key IN ({ph}) "
        "ORDER BY (i.rating > 0) DESC, i.rating DESC, COALESCE(agg.ai_score, 0) DESC, "
        "i.date_taken DESC, i.key DESC LIMIT 1"
    )
    row = db.execute(sql, burst_keys).fetchone()
    if not row:
        return None
    r = row.get("k") if isinstance(row, dict) else row[0]
    return str(r) if r is not None else None


def stack_metadata_for_api(db: sqlite3.Connection, stack_id: int) -> dict | None:
    """Stack row plus member keys; ``stack_member_count`` matches live membership."""
    row = db.execute(
        """
        SELECT stack_id, representative_key, stack_size
        FROM image_stacks
        WHERE stack_id = ?
        """,
        (stack_id,),
    ).fetchone()
    if not row:
        return None
    mem_rows = db.execute(
        """
        SELECT image_key FROM image_stack_members
        WHERE stack_id = ?
        ORDER BY image_key ASC
        """,
        (stack_id,),
    ).fetchall()
    member_keys = [str(r["image_key"]) for r in mem_rows]
    return {
        "stack_id": int(row["stack_id"]),
        "representative_key": str(row["representative_key"]),
        "stack_member_count": len(member_keys),
        "member_keys": member_keys,
    }


def catalog_image_stack_row_fields(db: sqlite3.Connection, image_key: str) -> dict:
    """Stack columns aligned with catalog list / by-keys rows for detail API parity.

    Non-members and solo images return ``stack_id`` / ``stack_member_count`` as
    ``None`` and ``is_stack_representative`` false.
    """
    row = db.execute(
        """
        SELECT st.stack_id AS stack_id, st.stack_size AS stack_member_count,
        CASE WHEN st.stack_id IS NOT NULL AND i.key = st.representative_key
             THEN 1 ELSE 0 END AS is_stack_representative
        FROM images i
        LEFT JOIN image_stack_members AS m_st ON m_st.image_key = i.key
        LEFT JOIN image_stacks AS st ON st.stack_id = m_st.stack_id
        WHERE i.key = ?
        """,
        (image_key,),
    ).fetchone()
    if not row or row["stack_id"] is None:
        return {
            "stack_id": None,
            "stack_member_count": None,
            "is_stack_representative": False,
        }
    smc = row["stack_member_count"]
    return {
        "stack_id": int(row["stack_id"]),
        "stack_member_count": int(smc) if smc is not None else None,
        "is_stack_representative": bool(row["is_stack_representative"]),
    }


def stack_split_member_out(
    db: sqlite3.Connection, stack_id: int, image_key: str
) -> dict:
    """Remove *image_key* from *stack_id*; dissolve singleton remnants to solo images.

    Call inside :func:`library_write` so the operation is one atomic transaction.

    Returns:
        ``split_out_key``, ``remaining_stack`` (metadata dict or ``None`` if dissolved),
        ``dissolved`` (True when no ``image_stacks`` row remains for former members).
    """
    stack_row = db.execute(
        "SELECT stack_id, representative_key FROM image_stacks WHERE stack_id = ?",
        (stack_id,),
    ).fetchone()
    if not stack_row:
        raise StackMutationError("stack not found", status_code=404)

    mem = db.execute(
        """
        SELECT 1 AS o FROM image_stack_members
        WHERE stack_id = ? AND image_key = ?
        """,
        (stack_id, image_key),
    ).fetchone()
    if not mem:
        raise StackMutationError(
            "image_key is not a member of this stack", status_code=400
        )

    db.execute(
        "DELETE FROM image_stack_members WHERE stack_id = ? AND image_key = ?",
        (stack_id, image_key),
    )

    remaining_rows = db.execute(
        """
        SELECT image_key FROM image_stack_members
        WHERE stack_id = ?
        ORDER BY image_key ASC
        """,
        (stack_id,),
    ).fetchall()
    remaining_keys = [str(r["image_key"]) for r in remaining_rows]

    if not remaining_keys:
        db.execute("DELETE FROM image_stacks WHERE stack_id = ?", (stack_id,))
        return {
            "split_out_key": image_key,
            "remaining_stack": None,
            "dissolved": True,
        }

    if len(remaining_keys) == 1:
        lone = remaining_keys[0]
        db.execute(
            "DELETE FROM image_stack_members WHERE stack_id = ? AND image_key = ?",
            (stack_id, lone),
        )
        db.execute("DELETE FROM image_stacks WHERE stack_id = ?", (stack_id,))
        return {
            "split_out_key": image_key,
            "remaining_stack": None,
            "dissolved": True,
        }

    old_rep = str(stack_row["representative_key"])
    new_rep = (
        old_rep
        if old_rep in remaining_keys
        else select_stack_representative_key_for_keys(db, remaining_keys)
    )
    if not new_rep or new_rep not in remaining_keys:
        new_rep = sorted(remaining_keys)[-1]

    n = len(remaining_keys)
    # stack_metadata_for_api is authoritative for API stack_member_count (live membership).
    db.execute(
        """
        UPDATE image_stacks
        SET representative_key = ?, stack_size = ?, user_modified = 1
        WHERE stack_id = ?
        """,
        (new_rep, n, stack_id),
    )
    meta = stack_metadata_for_api(db, stack_id)
    assert meta is not None
    return {
        "split_out_key": image_key,
        "remaining_stack": meta,
        "dissolved": False,
    }


def stack_merge_into(
    db: sqlite3.Connection, target_stack_id: int, source_stack_id: int
) -> dict:
    """Move all members from *source_stack_id* into *target_stack_id*; delete source stack.

    Call inside :func:`library_write`.

    Returns:
        ``stack`` (target metadata), ``merged_stack_id`` (source id).
    """
    if target_stack_id == source_stack_id:
        raise StackMutationError("cannot merge a stack into itself", status_code=400)

    t_row = db.execute(
        "SELECT stack_id, representative_key FROM image_stacks WHERE stack_id = ?",
        (target_stack_id,),
    ).fetchone()
    s_row = db.execute(
        "SELECT stack_id FROM image_stacks WHERE stack_id = ?",
        (source_stack_id,),
    ).fetchone()
    if not t_row or not s_row:
        raise StackMutationError("stack not found", status_code=404)

    db.execute(
        """
        UPDATE image_stack_members
        SET stack_id = ?
        WHERE stack_id = ?
        """,
        (target_stack_id, source_stack_id),
    )
    db.execute("DELETE FROM image_stacks WHERE stack_id = ?", (source_stack_id,))

    member_rows = db.execute(
        """
        SELECT image_key FROM image_stack_members
        WHERE stack_id = ?
        ORDER BY image_key ASC
        """,
        (target_stack_id,),
    ).fetchall()
    keys = [str(r["image_key"]) for r in member_rows]
    n = len(keys)
    if n == 0:
        db.execute("DELETE FROM image_stacks WHERE stack_id = ?", (target_stack_id,))
        raise StackMutationError("merge produced an empty stack", status_code=500)

    old_rep = str(t_row["representative_key"])
    new_rep = (
        old_rep
        if old_rep in keys
        else select_stack_representative_key_for_keys(db, keys)
    )
    if not new_rep or new_rep not in keys:
        new_rep = sorted(keys)[-1]

    # stack_metadata_for_api is authoritative for API stack_member_count (live membership).
    db.execute(
        """
        UPDATE image_stacks
        SET representative_key = ?, stack_size = ?, user_modified = 1
        WHERE stack_id = ?
        """,
        (new_rep, n, target_stack_id),
    )
    meta = stack_metadata_for_api(db, target_stack_id)
    assert meta is not None
    return {"stack": meta, "merged_stack_id": source_stack_id}


def stack_set_representative(
    db: sqlite3.Connection, stack_id: int, new_representative_key: str
) -> dict:
    """Set stack representative to *new_representative_key* (must be a member).

    Call inside :func:`library_write`.
    """
    stack_row = db.execute(
        "SELECT stack_id FROM image_stacks WHERE stack_id = ?",
        (stack_id,),
    ).fetchone()
    if not stack_row:
        raise StackMutationError("stack not found", status_code=404)

    mem = db.execute(
        """
        SELECT 1 AS o FROM image_stack_members
        WHERE stack_id = ? AND image_key = ?
        """,
        (stack_id, new_representative_key),
    ).fetchone()
    if not mem:
        raise StackMutationError(
            "image_key is not a member of this stack", status_code=400
        )

    db.execute(
        """
        UPDATE image_stacks
        SET representative_key = ?, user_modified = 1
        WHERE stack_id = ?
        """,
        (new_representative_key, stack_id),
    )
    # Keep stack_size aligned with membership
    cnt_row = db.execute(
        "SELECT COUNT(*) AS c FROM image_stack_members WHERE stack_id = ?",
        (stack_id,),
    ).fetchone()
    n = int(cnt_row["c"]) if cnt_row else 0
    # stack_metadata_for_api is authoritative for API stack_member_count (live membership).
    db.execute(
        "UPDATE image_stacks SET stack_size = ? WHERE stack_id = ?",
        (n, stack_id),
    )
    meta = stack_metadata_for_api(db, stack_id)
    assert meta is not None
    return {"stack": meta}

File: apps/visualizer/backend-ts/src/db/library/stacks.ts
/**
 * Burst stack membership and mutations. Port of `core/database/stacks.py`.
 *
 * Every mutation here must run inside `libraryWrite` so it is one atomic
 * transaction — a half-applied split leaves `image_stacks.stack_size` disagreeing
 * with `image_stack_members`, which the grid reads as a phantom stack.
 */
import type { Db } from '../connection.js';

/** Invalid stack edit; `statusCode` is the HTTP status the API maps it to. */
export class StackMutationError extends Error {
  readonly statusCode: number;

  constructor(message: string, statusCode = 400) {
    super(message);
    this.name = 'StackMutationError';
    this.statusCode = statusCode;
  }
}

export interface StackMetadata {
  stack_id: number;
  representative_key: string;
  stack_member_count: number;
  member_keys: string[];
}

/**
 * Lexicographically largest key, the last-resort representative.
 *
 * Python used `sorted(keys)[-1]`, which orders by Unicode code point; JavaScript's
 * default sort orders by UTF-16 code unit. The two differ only above the BMP, and
 * every key in this catalog is drawn from alphanumerics, `-`, `_`, space and
 * parentheses, so they agree here.
 */
function largestKey(keys: readonly string[]): string {
  return [...keys].sort()[keys.length - 1]!;
}

export function stackExists(db: Db, stackId: number): boolean {
  const row = db
    .prepare('SELECT 1 AS o FROM image_stacks WHERE stack_id = ? LIMIT 1')
    .get(stackId);
  return row !== undefined;
}

/** Member keys for `stackId`, ascending. */
export function listStackMemberKeys(db: Db, stackId: number): string[] {
  const rows = db
    .prepare(
      `
        SELECT image_key FROM image_stack_members
        WHERE stack_id = ?
        ORDER BY image_key ASC
        `,
    )
    .all(stackId) as { image_key: string }[];
  return rows.map((r) => String(r.image_key));
}

/**
 * All keys in the stack containing `catalogKey`, or `[catalogKey]` when it is a
 * solo image.
 */
export function listCatalogStackMemberKeys(db: Db, catalogKey: string): string[] {
  const row = db
    .prepare('SELECT stack_id FROM image_stack_members WHERE image_key = ? LIMIT 1')
    .get(catalogKey) as { stack_id: number } | undefined;
  if (!row) return [catalogKey];
  return listStackMemberKeys(db, Math.trunc(row.stack_id));
}

/**
 * Pick a representative using the same ranking as the `batch_stack_detect` job:
 * rated images first, then higher rating, then higher mean active-perspective
 * score, then newest, then largest key.
 */
export function selectStackRepresentativeKeyForKeys(
  db: Db,
  keys: readonly string[],
): string | null {
  const burstKeys = keys.filter(Boolean).map(String);
  if (burstKeys.length === 0) return null;
  const ph = burstKeys.map(() => '?').join(',');
  const row = db
    .prepare(
      'SELECT i.key AS k FROM images i ' +
        'LEFT JOIN ( ' +
        '  SELECT s.image_key AS image_key, AVG(s.score) AS ai_score ' +
        '  FROM image_scores s ' +
        '  INNER JOIN perspectives p ON p.slug = s.perspective_slug AND p.active = 1 ' +
        "  WHERE s.is_current = 1 AND s.image_type = 'catalog' " +
        '  GROUP BY s.image_key ' +
        ') agg ON agg.image_key = i.key ' +
        `WHERE i.key IN (${ph}) ` +
        'ORDER BY (i.rating > 0) DESC, i.rating DESC, COALESCE(agg.ai_score, 0) DESC, ' +
        'i.date_taken DESC, i.key DESC LIMIT 1',
    )
    .get(...burstKeys) as { k: string | null } | undefined;
  if (!row || row.k === null || row.k === undefined) return null;
  return String(row.k);
}

/**
 * Stack row plus member keys.
 *
 * `stack_member_count` comes from live membership rather than the stored
 * `stack_size`, which is why this is the authoritative shape for API responses.
 */
export function stackMetadataForApi(db: Db, stackId: number): StackMetadata | null {
  const row = db
    .prepare(
      `
        SELECT stack_id, representative_key, stack_size
        FROM image_stacks
        WHERE stack_id = ?
        `,
    )
    .get(stackId) as { stack_id: number; representative_key: string } | undefined;
  if (!row) return null;
  const memberKeys = listStackMemberKeys(db, stackId);
  return {
    stack_id: Math.trunc(row.stack_id),
    representative_key: String(row.representative_key),
    stack_member_count: memberKeys.length,
    member_keys: memberKeys,
  };
}

export interface StackRowFields {
  stack_id: number | null;
  stack_member_count: number | null;
  is_stack_representative: boolean;
}

/**
 * Stack columns aligned with the catalog list / by-keys rows, so the detail API
 * reports the same stack state the grid does.
 */
export function catalogImageStackRowFields(db: Db, imageKey: string): StackRowFields {
  const row = db
    .prepare(
      `
        SELECT st.stack_id AS stack_id, st.stack_size AS stack_member_count,
        CASE WHEN st.stack_id IS NOT NULL AND i.key = st.representative_key
             THEN 1 ELSE 0 END AS is_stack_representative
        FROM images i
        LEFT JOIN image_stack_members AS m_st ON m_st.image_key = i.key
        LEFT JOIN image_stacks AS st ON st.stack_id = m_st.stack_id
        WHERE i.key = ?
        `,
    )
    .get(imageKey) as
    | { stack_id: number | null; stack_member_count: number | null; is_stack_representative: number }
    | undefined;
  if (!row || row.stack_id === null) {
    return { stack_id: null, stack_member_count: null, is_stack_representative: false };
  }
  return {
    stack_id: Math.trunc(row.stack_id),
    stack_member_count: row.stack_member_count === null ? null : Math.trunc(row.stack_member_count),
    is_stack_representative: Boolean(row.is_stack_representative),
  };
}

export interface SplitMemberResult {
  split_out_key: string;
  remaining_stack: StackMetadata | null;
  dissolved: boolean;
}

/**
 * Remove `imageKey` from `stackId`, dissolving singleton remnants back to solo
 * images. Call inside `libraryWrite`.
 *
 * A one-member stack is not a stack, so removing the second-to-last member deletes
 * the stack row as well and reports `dissolved: true`.
 */
export function stackSplitMemberOut(
  db: Db,
  stackId: number,
  imageKey: string,
): SplitMemberResult {
  const stackRow = db
    .prepare('SELECT stack_id, representative_key FROM image_stacks WHERE stack_id = ?')
    .get(stackId) as { representative_key: string } | undefined;
  if (!stackRow) throw new StackMutationError('stack not found', 404);

  const mem = db
    .prepare('SELECT 1 AS o FROM image_stack_members WHERE stack_id = ? AND image_key = ?')
    .get(stackId, imageKey);
  if (!mem) throw new StackMutationError('image_key is not a member of this stack', 400);

  db.prepare('DELETE FROM image_stack_members WHERE stack_id = ? AND image_key = ?').run(
    stackId,
    imageKey,
  );

  const remainingKeys = listStackMemberKeys(db, stackId);

  if (remainingKeys.length === 0) {
    db.prepare('DELETE FROM image_stacks WHERE stack_id = ?').run(stackId);
    return { split_out_key: imageKey, remaining_stack: null, dissolved: true };
  }

  if (remainingKeys.length === 1) {
    db.prepare('DELETE FROM image_stack_members WHERE stack_id = ? AND image_key = ?').run(
      stackId,
      remainingKeys[0]!,
    );
    db.prepare('DELETE FROM image_stacks WHERE stack_id = ?').run(stackId);
    return { split_out_key: imageKey, remaining_stack: null, dissolved: true };
  }

  const oldRep = String(stackRow.representative_key);
  let newRep = remainingKeys.includes(oldRep)
    ? oldRep
    : selectStackRepresentativeKeyForKeys(db, remainingKeys);
  if (!newRep || !remainingKeys.includes(newRep)) newRep = largestKey(remainingKeys);

  db.prepare(
    `
        UPDATE image_stacks
        SET representative_key = ?, stack_size = ?, user_modified = 1
        WHERE stack_id = ?
        `,
  ).run(newRep, remainingKeys.length, stackId);

  const meta = stackMetadataForApi(db, stackId);
  if (!meta) throw new StackMutationError('stack disappeared during split', 500);
  return { split_out_key: imageKey, remaining_stack: meta, dissolved: false };
}

export interface MergeResult {
  stack: StackMetadata;
  merged_stack_id: number;
}

/**
 * Move every member of `sourceStackId` into `targetStackId` and delete the source
 * stack row. Call inside `libraryWrite`.
 */
export function stackMergeInto(
  db: Db,
  targetStackId: number,
  sourceStackId: number,
): MergeResult {
  if (targetStackId === sourceStackId) {
    throw new StackMutationError('cannot merge a stack into itself', 400);
  }

  const tRow = db
    .prepare('SELECT stack_id, representative_key FROM image_stacks WHERE stack_id = ?')
    .get(targetStackId) as { representative_key: string } | undefined;
  const sRow = db
    .prepare('SELECT stack_id FROM image_stacks WHERE stack_id = ?')
    .get(sourceStackId);
  if (!tRow || !sRow) throw new StackMutationError('stack not found', 404);

  db.prepare('UPDATE image_stack_members SET stack_id = ? WHERE stack_id = ?').run(
    targetStackId,
    sourceStackId,
  );
  db.prepare('DELETE FROM image_stacks WHERE stack_id = ?').run(sourceStackId);

  const keys = listStackMemberKeys(db, targetStackId);
  if (keys.length === 0) {
    db.prepare('DELETE FROM image_stacks WHERE stack_id = ?').run(targetStackId);
    throw new StackMutationError('merge produced an empty stack', 500);
  }

  const oldRep = String(tRow.representative_key);
  let newRep = keys.includes(oldRep) ? oldRep : selectStackRepresentativeKeyForKeys(db, keys);
  if (!newRep || !keys.includes(newRep)) newRep = largestKey(keys);

  db.prepare(
    `
        UPDATE image_stacks
        SET representative_key = ?, stack_size = ?, user_modified = 1
        WHERE stack_id = ?
        `,
  ).run(newRep, keys.length, targetStackId);

  const meta = stackMetadataForApi(db, targetStackId);
  if (!meta) throw new StackMutationError('stack disappeared during merge', 500);
  return { stack: meta, merged_stack_id: sourceStackId };
}

/** The stack containing `imageKey`, or `null` for a solo image. */
export function stackIdForImageKey(db: Db, imageKey: string): number | null {
  const row = db
    .prepare('SELECT stack_id FROM image_stack_members WHERE image_key = ? LIMIT 1')
    .get(imageKey) as { stack_id: number } | undefined;
  return row ? Math.trunc(row.stack_id) : null;
}

/**
 * Point `stackId` at a new representative, which must already be a member.
 * Call inside `libraryWrite`.
 */
export function stackSetRepresentative(
  db: Db,
  stackId: number,
  newRepresentativeKey: string,
): { stack: StackMetadata } {
  const stackRow = db
    .prepare('SELECT stack_id FROM image_stacks WHERE stack_id = ?')
    .get(stackId);
  if (!stackRow) throw new StackMutationError('stack not found', 404);

  const mem = db
    .prepare('SELECT 1 AS o FROM image_stack_members WHERE stack_id = ? AND image_key = ?')
    .get(stackId, newRepresentativeKey);
  if (!mem) throw new StackMutationError('image_key is not a member of this stack', 400);

  db.prepare(
    'UPDATE image_stacks SET representative_key = ?, user_modified = 1 WHERE stack_id = ?',
  ).run(newRepresentativeKey, stackId);

  // Realign the denormalized stack_size with actual membership while we hold the
  // writer seat; a stale value here shows a wrong badge count in the grid.
  const cntRow = db
    .prepare('SELECT COUNT(*) AS c FROM image_stack_members WHERE stack_id = ?')
    .get(stackId) as { c: number } | undefined;
  db.prepare('UPDATE image_stacks SET stack_size = ? WHERE stack_id = ?').run(
    cntRow ? Math.trunc(cntRow.c) : 0,
    stackId,
  );

  const meta = stackMetadataForApi(db, stackId);
  if (!meta) throw new StackMutationError('stack disappeared during update', 500);
  return { stack: meta };
}
